### PR TITLE
Issue #5229 - WebSocket documentation.

### LINKED ISSUE #5229: WebSocket documentation in Jetty 10
**Jetty version**
10.0.x

**Description**
Write the documentation for websocket in jetty-10.

This should be split into different sections for each API.

- Client
    - Jetty
    - Javax
    - Core

- Server
    - Jetty
    - Javax
    - Core

### PR DESCRIPTION
**Pull request by sbordet**
WebSocket server documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Comment by gregw** [comment]
just working out why PRs are not building atm...

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+[[pg-server-websocket-configure-filter]]
+==== Advanced `WebSocketUpgradeFilter` Configuration
+
+`JavaxWebSocketServletContainerInitializer` installs the `WebSocketUpgradeFilter` to handle HTTP requests that upgrade to WebSocket.
+
+Typically, the `WebSocketUpgradeFilter` is not present in the `web.xml` configuration, and therefore `JavaxWebSocketServletContainerInitializer` creates a new `WebSocketUpgradeFilter` and installs it _before_ any other Filter declared in `web.xml`, under the default name of `"org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter"` and with path mapping `/*`.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: Should you not mention the `JettyWebSocketServletContainerInitializer` here as well?

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+However, if the `WebSocketUpgradeFilter` is already present in `web.xml` under the default name, then `JavaxWebSocketServletContainerInitializer` will use that declared in `web.xml` instead of creating a new one.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: It must have a specific name for it to be recognised and prevent the default filter from being installed. This name should be the fully qualified class name of the upgrade filter: `org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter`.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+[NOTE]
+====
+When using the Jetty WebSocket APIs, WebSocket endpoints must always be explicitly configured.
+====
+
+There are two ways of configuring WebSocket endpoints when using the Jetty WebSocket APIs:
+
+* xref:pg-server-websocket-jetty-endpoints-container[Using `JettyWebSocketServerContainer`], which is very similar to how WebSocket endpoints are configured using the standard `javax.websocket` APIs (see xref:pg-server-websocket-standard-endpoints[here]).
+* xref:pg-server-websocket-jetty-endpoints-servlet[Using `JettyWebSocketServlet`], which may be simpler to configure.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc]: Is this really simpler to configure?

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+An alternative way to register WebSocket endpoints using the Jetty WebSocket APIs is to use a `JettyWebSocketServlet` subclass.
+
+This method has the advantage that it does not install the `WebSocketUpgradeFilter` under the hoods, because the WebSocket upgrade is handled directly by your `JettyWebSocketServlet` subclass.
+
+This may also have a performance benefit for non-WebSocket HTTP requests (as they will not pass through the `WebSocketUpgradeFilter`), and be simpler to configure (for example, it is easier to configure the `CrossOriginFilter`, see also xref:pg-server-websocket-configure-filter[this section] for more information).
+
+[IMPORTANT]
+====
+Do not mix the use of `JettyWebSocketServerContainer` with the use of `JettyWebSocketServlet`.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc]: Probably worth leaving this warning as is because this would be inefficient, but you can actually use both the UpgradeServlet and UpgradeFilter together without any problem, they do not share their mappings. If the filter does not upgrade the request it may still go to the servlet and have it upgraded after.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc [hunk 5]
```diff
@@ -12,6 +12,51 @@
...
 [[pg-server-websocket]]
-=== WebSocket Server Libraries
+=== WebSocket Server
 
-TODO
+// JettyWebSocketServletContainerInitializer.configure(ServletContextHandler context, Configurator configurator); is one entry point.
+
+// JettyWebSocketServlet.configure(JettyWebSocketServletFactory factory); another entry point.
+
+// TODO: why they expose different interfaces? Configurator exposes JettyWebSocketServerContainer, the servlet exposes JettyWebSocketServletFactory => historical, live with it.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc]: This looks incomplete, TODOs and comments here.

File: jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyClientClassLoaderTest.java [hunk 6]
```diff
@@ -149,6 +149,11 @@ protected void configure(JettyWebSocketServletFactory factory)
     @WebSocket
     public static class EchoSocket
     {
+        public EchoSocket()
+        {
+            System.err.println("this = " + this);
+        }
+
```
Comment (lachlan-roberts) [inline jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyClientClassLoaderTest.java]: Revert this change.

File: jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyClientClassLoaderTest.java [hunk 7]
```diff
@@ -191,7 +196,7 @@ public void websocketProvidedByServer() throws Exception
             WebAppTester.WebApp app2 = webAppTester.createWebApp("/echo");
             app2.addConfiguration(new JettyWebSocketConfiguration());
             app2.createWebInf();
-            app2.copyClass(EchoServlet.class);
+//            app2.copyClass(EchoServlet.class);
```
Comment (lachlan-roberts) [inline jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyClientClassLoaderTest.java]: Revert this change.

**Review by lachlan-roberts — CHANGES_REQUESTED** [review]
You might also want to write about the `JettyWebSocketCreator` class which can be registered to pathspec and can be used to dynamically create the endpoint using information from the request and response.

Also maybe some information about the `JettyWebSocketServerContainer.upgrade()` feature which allows an upgrade directly without registering a mapping.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc (hunk 1 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: Done.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc (hunk 2 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: This is already mentioned in the previous paragraph.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc (hunk 3 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc]: Yes, in `web.xml`. I clarified that.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc (hunk 4 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-jetty.adoc]: Good point, I reworked the admonition and added a paragraph about using both.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc (hunk 5 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc]: Ah yes, they are leftovers. I removed them.

File: jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyClientClassLoaderTest.java (hunk 7 above)
Comment (sbordet) [reply jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/JettyClientClassLoaderTest.java]: Done.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+[[pg-server-websocket-configure-filter]]
+==== Advanced `WebSocketUpgradeFilter` Configuration
+
+The WebSocket ``ServletContainerInitializer``s, namely `JavaxWebSocketServletContainerInitializer` and `JettyWebSocketServletContainerInitializer`, install the `WebSocketUpgradeFilter` to handle HTTP requests that upgrade to WebSocket.
```
Comment (gregw) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: This is not correct. 

The jetty SCI doesn't add the filter, rather it is added in the addMapping method on the container. Only the javax SCI always adds the filter.

Moreover,  the SCIs do a lot more than just install the filter:  pools for inflaters, deflaters, buffers; executors; object factory; mappings; contains; lifecycle listeners etc.   

 The SCIs initialize a lot of machinery and much of it is extensible/configurable to the users so probably should be of note in the "Advanced" documentation.    So I think we should open with a full enumeration of the mechanism that need to be initialized and how/where they are.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+Typically, the `WebSocketUpgradeFilter` is not present in the `web.xml` configuration, and therefore the WebSocket ``ServletContainerInitializer``s create a new `WebSocketUpgradeFilter` and install it _before_ any other Filter declared in `web.xml`, under the default name of `"org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter"` and with path mapping `/*`.
```
Comment (gregw) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: Once we have described all the bits that are initialized, then we need subsections saying how each bit of machinery can be customized.  The filter being only one of these subsections.  We also need to say how pools etc. can be customized/configured/replaced.

**Review by gregw — CHANGES_REQUESTED** [review]
I only looked at the first bit and think it needs a bit of refactoring. 
I'll look deeper once @lachlan-roberts has commented.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc (hunk 9 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-filter.adoc]: @gregw the machinery can be customized, but without APIs. 
It's all relying on some magic "look at this attribute in the context" and the attribute names are based on classes that are not really meant to be exposed in this way, say for example `o.e.j.w.core.WebSocketComponents`.
I would not document this now, and if people really struggle, then add the docs later.

**Comment by sbordet** [comment]
@lachlan-roberts @gregw I have updated the documentation, see also my comments about your comments.
Please re-review.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc [hunk 10]
```diff
@@ -12,6 +12,33 @@
...
+
+The standard APIs provide few features that are not present in the Jetty WebSocket APIs:
+
+* Encoders and Decoders for automatic conversion of text or binary messages to objects.
+* `Reader` and `InputStream` for simple, blocking, message streaming.
+* Simple URI template matching.
+
+On the other hand, the Jetty WebSocket APIs are more efficient and offer greater and more fine-grained control, and provide features that are not present in the standard APIs:
+
+* `MessageSink` for advanced, asynchronous, message streaming with backpressure.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc]: `MessageSink` is a websocket-core internal interface. It's not one of the features of the Jetty WebSocket API.

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-standard.adoc [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+[NOTE]
+====
+The `javax.websocket-api` artifact and the `websocket-javax-server` artifact (and its transitive dependencies) should be present in the server class-path (or module-path), and never in the web application's `/WEB-INF/lib` directory.
+====
+
+To configure correctly your WebSocket application based on the standard `javax.websocket` APIs, you need two steps:
+
+. Make sure that Jetty xref:pg-server-websocket-standard-container[sets up] an instance of `javax.websocket.server.ServerContainer`.
+. Use the `ServerContainer` APIs in your applications to xref:pg-server-websocket-standard-endpoints[register your WebSocket endpoints] that implement your application logic.
```
Comment (lachlan-roberts) [inline documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket-standard.adoc]: I know you mention this later but you don't have to use the `ServerContainer` APIs, you could also use the `javax.websocket.server.ServerEndpoint` annotation to register the endpoints.

**Review by lachlan-roberts — CHANGES_REQUESTED** [review]

File: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc (hunk 10 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/websocket/server-websocket.adoc]: Removed.

**Review by gregw — APPROVED** [review]

**Review by lachlan-roberts — APPROVED** [review]
